designs/cloudstack-metrics-bind-addr.md

```diff
@@ -0,0 +1,176 @@
+parameter defaults to *localhost:8080*, making the port only visible to other
+containers in the manager's pod (sidecars).  To expose this port from the pod
+one needs to bind it to the pod's NIC interface, typically by setting it to
+*0.0.0.0:8080* or just *:8080* (since the actual IP address that will be assigned to the pod is not easily known).
+
+CAPC users can easily add this argument to an overridden infrastructure-components.yaml.
+EKS-A users, however, have no available mechanism for doing so to the
+infrastructure-components.yaml that EKS-A will use.
+
+This design proposes a simple mechanism for supporting this.  The scope of the proposed change is CAPC,
+though CAPV, CAPD and CAPT are known to support the same parameter (and other providers likely do too.
+
+### Tenets
+
+* ****Simple:**** simple to use, simple to understand, simple to maintain
+* ****Minimally Impactful:**** to the EKS-A code base.
+* ****Secure by Default:**** if unused defaults to the most secure option
+
+### Goals and Objectives
+
+As a Kubernetes administrator I want to be able to expose the CAPC Manager's
+metrics port from the pod running it so I can scrape metrics from it.
+
+### Statement of Scope
+
+**In scope**
+
+* Add support for overriding the default CAPC manager's metrics-bind-addr at
+cluster provisioning time.**
+
+**Not in scope**
+
+* Doing this for all EKS-A providers.
+
+**Future scope**
+
+* Implement this for the other EKS-A providers.
+
+
+## Overview of Solution
+
+We propose to
+* Modify the CAPC infrastructure-components.yaml to define argument *--metrics-bind-addr* for the CAPC Manager deployment, from an optional environment variable which defaults to the same setting as the CAPC manager uses for this parameter.
+
+* Modify function EnvMap() in cloudstack.go, which sets up the environment that *clusterctl init* will run under, to read the value of a new environment
+variable (if defined by the provisioner in their *eksctl anywhere create cluster* session) into the map that it creates.
+
+## Solution Details
+
+### CAPC infrastructure-components.yaml change
+
+The deployment spec for the CAPC manager will be modified to define a *--metrics-bind-address*
+for the deployment's pod template's manager container, resolved with an environment
+variable substitution defaulting to the original secure default used by the manager:
+
+```
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+...
+  name: capc-controller-manager
+  namespace: capc-system
+spec:
+...
+    spec:
+      containers:
+      - name: manager
+        args:
+        - --leader-elect
+        - --metrics-bind-addr=${CAPC_MANAGER_METRICS_BIND_ADDR:-localhost:8080}
+        command:
+        - /manager
+        image: localhost:5000/cluster-api-provider-cloudstack:v0.4.5
+...
+```
+
+### EKS-A change to expose this environment variable, if defined in the provisioner's environment at the time of *eksctl anywhere cluster create*:
```

Comment:
can we talk a bit about upgrade case? For example, if user specify this env during create, and not specify it during upgrade, will the capas pod be recreated with default metrics address? Also if `CloudStackMetricsBindAddr` is the only thing changed when running upgrade cmd, how do we trigger the components upgrade run.

Reply:
Agreed Joey, great point.  I think that makes the case for formally including this parameter in one of the EKS-A CRDs so it gets preserved at creation time for reuse in subsequent activities.  Probably CloudStackDataCenter unless there was interest in adopting this across all the providers.

Reply:
After reading your further comments @g-gaston, I think I'm understanding your point being that use of CAPI providers is an implementation detail, and you want to leave the door open to EKS-A using other kinds of providers in the future.

If so, fair enough.  Still, there is a need demonstrated by at least one key stakeholder for configuration of CAPC, so I think it serves the product to allow for this for this specific provider (and perhaps the other EKS-A CAPI providers individually as required).  Therefore, I'll re-propose that this new attribute go in the CloudStackDataCenter CRD, as that's the current home for all CAPC-specific configuration.